vm/binding/lib.rs

```diff
@@ -80,17 +104,114 @@ impl Default for VM {
 impl Drop for VM {
     fn drop(&mut self) {
         if cfg!(not(miri)) {
-            unsafe { moshell_vm_free(self.0) }
+            unsafe { moshell_vm_free(self.ffi) }
         }
     }
 }
 
 #[repr(C)]
 #[derive(Copy, Clone)]
-struct VmFFI(*mut ffi::c_void);
+struct VmArrayFFI(u64, *mut VmValueFFI);
+#[repr(C)]
+#[derive(Copy, Clone)]
+struct VmStringFFI(u64, *mut ffi::c_char);
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub union VmValueFFI {
```

Comment:
Keep this struct internal and promote the use of the already present safe alternative would be great.

If unnecessary conversions between the two interface are an issue, maybe treat the Rust part a simple safe wrapper with lifetimes?

Reply:
You need to know the kind of value held by the VmValuFFI transform it to a safe VmValue, however the VM can't know what's behind a VmValueFFI si it can only return an unsafe holder to a value. It's up to the user to use its convert methods safely

Reply:
Exposing an unsafe type and a safe type together without conversion functions, even unsafe ones, is bizarre. The point here is that this would require exposing the `analyzer` crate at the `dependencies` level, which is effectively meh.